File: 프로그래머스/lv2/42860. 조이스틱/조이스틱.cpp
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

int solution(string name) {
    int answer = 0;
    int len = name.length();
    int moveCnt = len - 1;
    
    for(int i = 0; i < len; i++) {
        answer += min(name[i] - 'A', 'Z' - name[i] + 1);
        
        int idx = i + 1;
        while(idx < len && name[idx] == 'A') idx++;
        
        moveCnt = min(moveCnt, i + len - idx + min(i, len - idx));
    }
    
    answer += moveCnt;
    
    return answer;
}

File: 프로그래머스/2/42860. 조이스틱/조이스틱.py
def solution(name):
    answer = 0
    length = len(name)
    min_move = length - 1
    while min_move >= 0 and name[min_move] == 'A':
        min_move -= 1
        
    if min_move < 0:
        return 0
        
    for i, c in enumerate(name):
        answer += min(ord(c) - ord('A'), ord('Z') - ord(c) + 1)
        
        end_A = i + 1
        while end_A < length and name[end_A] == 'A':
            end_A += 1
            
        min_move = min([min_move, i * 2 + (length - end_A), i + (length - end_A) * 2])
    
    answer += min_move    
    
    return answer